core/src/main/java/hudson/util/AtomicFileWriter.java

```diff
@@ -58,27 +64,51 @@ public AtomicFileWriter(File f) throws IOException {
     }
 
     /**
-     * @param encoding
-     *      File encoding to write. If null, platform default encoding is chosen.
+     * @param encoding File encoding to write. If null, platform default encoding is chosen.
+     *
+     * @deprecated Use {@link #AtomicFileWriter(Path, Charset)}
      */
-    public AtomicFileWriter(File f, String encoding) throws IOException {
-        File dir = f.getParentFile();
-        try {
-            dir.mkdirs();
-            tmpFile = File.createTempFile("atomic",null, dir);
-        } catch (IOException e) {
-            throw new IOException("Failed to create a temporary file in "+ dir,e);
-        }
-        destFile = f;
-        if (encoding==null)
-            encoding = Charset.defaultCharset().name();
+    @Deprecated
+    public AtomicFileWriter(@Nonnull File f, @Nullable String encoding) throws IOException {
+        this(toPath(f), encoding == null ? Charset.defaultCharset() : Charset.forName(encoding));
```

Comment:
Not sure why I only noticed this now but 
`Charset.forName()` throws
* `IllegalCharsetNameException` - If the given charset name is illegal
* `IllegalArgumentException` - If the given charsetName is null
* `UnsupportedCharsetException` - If no support for the named charset is available in this instance of the Java virtual machine

Reply:
Yes, should have added a comment here myself preempively: the previous code path was going through `core = new BufferedWriter(new OutputStreamWriter(Files.newOutputStream(tmpFile.toPath()), encoding));`

And `new OutputStreamWriter(Files.newOutputStream(tmpFile.toPath()), encoding)` ends up calling Charset.forName() already.

```java
public OutputStreamWriter(OutputStream out, String charsetName)
        throws UnsupportedEncodingException
    {
        super(out);
        if (charsetName == null)
            throw new NullPointerException("charsetName");
        se = StreamEncoder.forOutputStreamWriter(out, this, charsetName);
    }
```

which calls:

```java
public static StreamEncoder forOutputStreamWriter(OutputStream var0, Object var1, String var2) throws UnsupportedEncodingException {
        String var3 = var2;
        if (var2 == null) {
            var3 = Charset.defaultCharset().name();
        }

        try {
            if (Charset.isSupported(var3)) {
                return new StreamEncoder(var0, var1, Charset.forName(var3));
            }
        } catch (IllegalCharsetNameException var5) {
            ;
        }

        throw new UnsupportedEncodingException(var3);
    }
```

Granted, **if** the caller put a non existing encoding, they would previously get an `UnsupportedEncodingException` because then `Charset.isSupported(var3)` may have returned false.

But, I don't imagine anyone catching `UnsupportedEncodingException` to handle it instead of just using an existing encoding, and hence would be hit by such change?

Reply:
OK!